### PR TITLE
CNAME responses can occur with "Type: A" dns_sd_config requests

### PR DESCRIPTION
**Pull request by mattberther**

### CODE REVIEW

**Comment by brian-brazil** [comment]
The warning is correct, per RFC2782 a CNAME is invalid as a CNAME is an alias. Given that you have an invalid SRV record I'd suggest that you fix that, rather than trying to suppress a correct warning for everyone.

**Comment by mattberther** [comment]
Thanks for the comment, @brian-brazil. I wasn't intending to use a SRV record. I was receiving the warnings when doing DNS service discovery using "Type: A" - which I understand may not be the way this was intended to work. However, when specifying this type, I kept seeing the messages.

In this scenario, is your recommendation still to create an SRV record instead of the CNAME and then set up my service discovery to use Type: Srv?

**Comment by roidelapluie** [comment]
I think that the same DNS query should not return both CNAME and A records. 

See https://tools.ietf.org/html/rfc1912

> A CNAME record is not allowed to coexist with any other data.

**Comment by mattberther** [comment]
@roidelapluie It appears as though `msg.SetQuestion(dns.Fqdn("feeds.berther.io"), dns.TypeA)` follows the CNAME to get the actual IP address.

**Comment by brian-brazil** [comment]
Can you provide more details of exactly what your setup is, including Prometheus config and the relevant dig output?

**Comment by mattberther** [comment]
@brian-brazil I'm trying to use blackbox-exporter to do a minimal check against a prometheus job that I've configured using `dns_sd_configs`. I acknowledge that for this simple use case, I could very easily just use a static list of targets. However, I put this together as the smallest, reproducible scenario.

Also, another option might be to support `type: CNAME` in `dns_sd_configs`, but chasing the CNAME may be much more difficult to implement, when the current `type: A` ultimately gets the right IP address. It's just the warnings that I'm trying to reduce.

Configuration details are below. When I run this config, I get a the following warnings in my logs (about every 30 seconds):

```
level=warn ts=2020-11-23T16:38:58.463Z caller=dns.go:215 component="discovery manager scrape" discovery=dns msg="Invalid SRV record" record="feeds.berther.io.\t300\tIN\tCNAME\t1vlhxd.feedproxy.ghs.google.com."
```

## Blackbox Exporter module

```yaml
modules:
  http_2xx:
    prober: http
    timeout: 5s
    http:
      method: GET
      fail_if_not_ssl: true
```

## Prometheus config

```yaml
- job_name: feeds.berther.io
  metrics_path: /probe
  params: {module: [http_2xx]}
  dns_sd_configs:
    - {names: [feeds.berther.io], type: A, port: 443}
  relabel_configs:
    - {source_labels: [__address__], target_label: __param_target, regex: (.*), replacement: "https://${1}"}
    - {source_labels: [__meta_dns_name], target_label: instance, regex: (.*), replacement: "https://${1}"}
    - {target_label: __address__, replacement: blackbox-exporter:9115}
```

## dig output

```sh
❯ dig feeds.berther.io

; <<>> DiG 9.10.6 <<>> feeds.berther.io
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 25814
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;feeds.berther.io.		IN	A

;; ANSWER SECTION:
feeds.berther.io.	300	IN	CNAME	1vlhxd.feedproxy.ghs.google.com.
1vlhxd.feedproxy.ghs.google.com. 300 IN	A	74.125.195.121

;; Query time: 158 msec
;; SERVER: 192.168.1.1#53(192.168.1.1)
;; WHEN: Mon Nov 23 16:38:37 MST 2020
;; MSG SIZE  rcvd: 106
```

**Comment by brian-brazil** [comment]
Okay, so that setup and response is fine. It looks like the issue is that we're not ignoring all the cnames on the way to the final answer - which is fine as long as it's not the target of a SRV record. In fact we never try in this code to resolve what's in the SRV record - that's done at scrape time. So this PR is correct but needs a more correct comment, and the log message after is also incorrect.

**Comment by mattberther** [comment]
I'm happy to make the corresponding updates, but am not clear what would be a "better comment" and what would make the log message "correct". 

Would the PR message be better stated as "allow CNAME record response to pass through" and the log message as simply "Invalid record" (removing the SRV text)?

**Comment by brian-brazil** [comment]
The current comment merely describes the code, so doesn't add anything. A more useful comment would indicate why it's okay to ignore a CNAME.

**Comment by mattberther** [comment]
@brian-brazil I've made the proposed changes. There seems to be a circleci test that is failing. I'm not expecting that my change of an error message caused the described failure (since the pipeline passed on the initial PR):

```
Failed
=== RUN   TestHandleMultipleQuitRequests
    web_test.go:492: 
        	Error Trace:	web_test.go:492
        	            				asm_amd64.s:1374
        	Error:      	Received unexpected error:
        	            	Post "http://localhost:9090/-/quit": EOF
        	Test:       	TestHandleMultipleQuitRequests
--- FAIL: TestHandleMultipleQuitRequests (5.01s)
```

However, I see no way of being able to re-run the workflow (presumably because i do not have write access). Is this something that you can kick off, or is there another way for me to re-run the workflow?

**Comment by brian-brazil** [comment]
I've kicked it off.

File: discovery/dns/dns.go [hunk 1]
```diff
@@ -211,8 +211,10 @@ func (d *Discovery) refreshOne(ctx context.Context, name string, ch chan<- *targ
 			target = hostPort(addr.A.String(), d.port)
 		case *dns.AAAA:
 			target = hostPort(addr.AAAA.String(), d.port)
+		case *dns.CNAME:
+			// Ignore to prevent warning message from default case.
```
Comment (brian-brazil) [inline discovery/dns/dns.go:215]: This comment is still not useful. I can already tell this from the code alone, instead explain why this is the right thing to do.

Comment (mattberther) [reply discovery/dns/dns.go:215]: I misunderstood what you meant by comment- I thought you were referring to the title of the PR. My mistake, I'll get it adjusted.

**Comment by mattberther** [comment]
Looks like all tests passed, and the proper text was updated, @brian-brazil. Please let me know if there's anything further needed on this PR.

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**